app/src/gui/components/map/tile-source.ts

```diff
@@ -0,0 +1,588 @@
+      if (tileRecord) {
+        const tileSetNames = tileRecord.sets;
+        if (tileSetNames.length === 1) {
+          await this.tileDB.delete(tileKey);
+        } else {
+          // remove the tile set name from the tile record
+          tileSetNames.splice(tileSetNames.indexOf(setName), 1);
+          await this.tileDB.put(tileRecord);
+        }
+      }
+    }
+  }
+}
+
+export class ImageTileStore extends TileStoreBase {
+  declare source: ImageTileSource;
+  declare tileLayer: TileLayer;
+
+  constructor() {
+    super();
+    this.source = new ImageTileSource({
+      attributions: ATTRIBUTION,
+      loader: this.tileLoader.bind(this),
+    });
+    this.tileLayer = new TileLayer({source: this.source});
+  }
+
+  getTileURL(): string | undefined {
+    return TILE_URL_MAP[MAP_SOURCE]['image'];
+  }
+
+  getTileGrid() {
+    return this.source.getTileGrid();
+  }
+
+  getTileLayer() {
+    return this.tileLayer;
+  }
+
+  getAttribution() {
+    return '&copy; OSM contributors';
+  }
+
+  async getTileAsDataURL(
+    z: number,
+    x: number,
+    y: number
+  ): Promise<string | null> {
+    const image = await this.getTileBlob(this.getURLForTile({z, x, y}));
+    return new Promise((resolve, reject) => {
+      if (!image) {
+        resolve(null);
+      } else {
+        const reader = new FileReader();
+        reader.onloadend = () => {
+          resolve(reader.result as string);
+        };
+        reader.onerror = reject;
+        reader.readAsDataURL(image);
+      }
+    });
+  }
+
+  /**
+   * @param {number} z The tile z coordinate.
+   * @param {number} x The tile x coordinate.
+   * @param {number} y The tile y coordinate.
+   * @param {LoaderOptions} options The loader options.
+   * @return {Promise<HTMLImageElement>} Resolves with a loaded image.
+   */
+  async tileLoader(
+    z: number,
+    x: number,
+    y: number,
+    options: LoaderOptions
+  ): Promise<HTMLImageElement> {
+    const image = new Image();
+    image.crossOrigin = options.crossOrigin ?? null;
+    image.src = (await this.getTileAsDataURL(z, x, y)) || '';
+    return image;
+  }
+}
+
+// A vector tile source
+
+export class VectorTileStore extends TileStoreBase {
+  declare source: VectorTileSource;
+  declare tileLayer: VectorTileLayer;
+
+  constructor() {
+    super();
+    this.source = new VectorTileSource({
+      attributions: ATTRIBUTION,
+      url: this.getTileURL(),
+      format: new MVT(),
+      maxZoom: 14,
+      tileLoadFunction: this.tileLoader.bind(this),
+    });
+
+    this.tileLayer = new VectorTileLayer({
+      source: this.source,
+      background: 'hsl(40, 26%, 93%)',
+    });
+    applyStyle(this.tileLayer, getMapStylesheet(MAP_STYLE), {
+      transformRequest: (url: string) => {
+        // TODO: cache these requests...
+        //console.log('transformRequest', url.replace('{key}', MAP_SOURCE_KEY));
+        return url.replace('{key}', MAP_SOURCE_KEY);
+      },
+    });
```

Comment:
Not sure how this relates to what you were describing re: using JSON stylesheet?

Reply:
i.e. do we still need to do funky caching of the style requests